### PR TITLE
rbd-fuse: librados will filter out -r option from command-line

### PR DESCRIPTION
**Pull request by maydaycc**
The -r option will be filtered out by librados
when exec cmd "rbd-fuse /mountpoint -p pool_name -r rbd_name"
other rbds can be seen under the mount point

Fixes: https://tracker.ceph.com/issues/56387
Signed-off-by: wanwencong <wanwc@chinatelecom.cn>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by maydaycc** [comment]
@tchaikov Would you please have a look at this PR?

File: src/rbd_fuse/rbd-fuse.cc [hunk 1]
```diff
@@ -912,20 +912,25 @@ connect_to_cluster(rados_t *pcluster)
...
-	std::map<std::string, bool> filter_args = {
-		{"-f", false},
-		{"-d", false}};
+	// librados will filter out -r/-f/-d options from command-line
+	std::map<std::string, int> filter_options = {
+		{"-r", OPT_NOT_FOUND},
+		{"-f", OPT_NOT_FOUND},
+		{"-d", OPT_NOT_FOUND}};
+
+	std::set<std::string> require_arg_options = { "-r" };
```
Comment (idryomov) [inline src/rbd_fuse/rbd-fuse.cc:927]:
```suggestion
	std::set<std::string> require_arg_options = {"-r"};
```

File: src/rbd_fuse/rbd-fuse.cc [hunk 2]
```diff
@@ -941,9 +946,14 @@ int main(int argc, const char *argv[])
 		exit(1);
 	}
 
-	for (auto& it : filter_args) {
-		if (it.second) {
+	for (auto& it : filter_options) {
+		if (it.second != OPT_NOT_FOUND) {
 			arg_vector.push_back(it.first.c_str());
+			if (require_arg_options.find(it.first.c_str()) == require_arg_options.end()) {
```
Comment (idryomov) [inline src/rbd_fuse/rbd-fuse.cc:952]: `require_arg_options` stores `std::string` so no need for `c_str`:
```suggestion
			if (require_arg_options.find(it.first) == require_arg_options.end()) {
```

File: src/rbd_fuse/rbd-fuse.cc [hunk 3]
```diff
@@ -941,9 +946,14 @@ int main(int argc, const char *argv[])
...
+				continue;
+			}else if (it.second + 1 < argc) {
```
Comment (idryomov) [inline src/rbd_fuse/rbd-fuse.cc:954]: Combine these two conditions and get rid of `continue`:
```
if (require_arg_options.find(it.first) != require_arg_options.end() &&
    it.second + 1 < argc) {
        arg_vector.push_back(argv[it.second + 1]);
}
```

**Review by idryomov — CHANGES_REQUESTED** [review]

**Comment by idryomov** [comment]
@maydaycc Are you actually using `rbd-fuse`?  If so, may I ask why?

`rbd-nbd` has been the preferred solution for many years now, is there something missing there that made you resort to `rbd-fuse`?

**Comment by idryomov** [comment]
Looks like this was missed in https://github.com/ceph/ceph/pull/20598, I wonder why -- `-r`/`--client_mountpoint` option is very old options (~2011).

**Comment by maydaycc** [comment]
> @maydaycc Are you actually using `rbd-fuse`? If so, may I ask why?
> 
> `rbd-nbd` has been the preferred solution for many years now, is there something missing there that made you resort to `rbd-fuse`?

In the process of creating volumes from images, we use rbd-fuse to mount rbd to the cinder volume node, and then use the resize2fs tool to adjust the size of the file system. rbd-nbd is a better choice but our kernel does not support the nbd module.

File: src/rbd_fuse/rbd-fuse.cc [hunk 4]
```diff
@@ -941,9 +946,13 @@ int main(int argc, const char *argv[])
 		exit(1);
 	}
 
-	for (auto& it : filter_args) {
-		if (it.second) {
+	for (auto& it : filter_options) {
+		if (it.second != OPT_NOT_FOUND) {
 			arg_vector.push_back(it.first.c_str());
+			if (require_arg_options.find(it.first) != require_arg_options.end() &&
```
Comment (trociny) [inline src/rbd_fuse/rbd-fuse.cc:952]: nit: you could use `require_arg_options.count(it.first)` to make it shorter.

Comment (maydaycc) [reply src/rbd_fuse/rbd-fuse.cc:952]: Thank you for your suggestion

File: src/rbd_fuse/rbd-fuse.cc [hunk 5]
```diff
@@ -941,9 +946,13 @@ int main(int argc, const char *argv[])
...
 
-	for (auto& it : filter_args) {
-		if (it.second) {
+	for (auto& it : filter_options) {
+		if (it.second != OPT_NOT_FOUND) {
 			arg_vector.push_back(it.first.c_str());
+			if (require_arg_options.count(it.first) &&
+			    it.second + 1 < argc) {
+				arg_vector.push_back(argv[it.second + 1]);
+			}
```
Comment (idryomov) [inline src/rbd_fuse/rbd-fuse.cc:955]: stray whitespace
```suggestion
			}
```

**Review by idryomov — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by idryomov** [comment]
@maydaycc Please fold the whitespace fixup into the original commit.

**Comment by maydaycc** [comment]
> @maydaycc Please fold the whitespace fixup into the original commit.

Done!

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
The only job in the suite that happens to exercise `rbd-fuse`:

https://pulpito.ceph.com/dis-2022-07-04_09:14:19-rbd-wip-dis-testing-distro-default-smithi/

**Pull request merged**